src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java

```diff
@@ -80,11 +84,60 @@
                 out.add(new Tuple2<>(pair.key(header), pair));
             }
             return out;
-        }).groupByKey();
+        }).groupByKey(parallelism);
 
         return markPairedEnds(keyedPairs, scoringStrategy, finder, header);
     }
 
+    /**
+     * Like <code>groupByKey</code>, but assumes that values are already sorted by key, so no shuffle is needed,
```

Comment:
Yes, the reads in a pair should always be in the same read group. The issue is one of namespace: read names not guaranteed to be unique across multiple read groups, and so we have to account for the possibility of a pair with name N in read group G, and another, different pair also with name N in read group G2. With the read names typically used at, eg., the Broad, this should not happen in practice, since the names include information like timestamps and should be globally unique, but others are free to use names that are not globally unique.